### PR TITLE
Remove some reflection usage

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
We've been using reflection as a way to abstract over Java API versions. Now that we're based on Java 8, some of this reflection usage is no longer necessary. In some cases, the APIs we were reflecting on are now directly available. In other cases, we'd like a little more performance and can call through method handles instead.

Modification:
Remove some reflection usage that was necessary when running on Java 6 or 7, and replace it with direct calls if the code is still needed. Replace the more performance sensitive reflection usage with MethodHandles.

Also remove the animal sniffer maven plug-in, because animal sniffer does not support signature polymorphic method calls.
See https://github.com/mojohaus/animal-sniffer/issues/67

Result:
Cleaner, and perhaps slightly faster, code.

Fixes https://github.com/netty/netty/issues/14009

### CODE REVIEW

**Comment by chrisvest** [comment]
Not all reflection usages have been replaced, because 1) reflect gives access to annotations, and method handles don't, and 2) some cases are just not performance sensitive so it's not worth the effort to change code that works.

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java [hunk 1]
```diff
@@ -65,41 +80,24 @@ public Object run() {
...
-                fieldOffset = PlatformDependent0.objectFieldOffset(cleanerField);
-                cleaner = PlatformDependent0.getObject(direct, fieldOffset);
-            } else {
-                fieldOffset = -1;
-                cleaner = cleanerField.get(direct);
-            }
-            clean = cleaner.getClass().getDeclaredMethod("clean");
-            clean.invoke(cleaner);
+            clean = (MethodHandle) mayBeCleanerField;
+            clean.invokeExact(direct);
```
Comment (franz1981) [inline common/src/main/java/io/netty/util/internal/CleanerJava6.java:84]: I've never played with method handle like this (always used `invokeExact` on unreflected stuff) - let's check the benchmarks to see if it is as  expected!

**Review by franz1981 — COMMENTED** [review]
Need to give another round but overall look fine!

**Comment by franz1981** [comment]
@chrisvest The allocate uninitialized arrays method, if uses `Methodhandle::invokeExact` doesn't need to have *any* limit: it will just go faster in any case.
The limit in bytes was because reflective calls over intrinsic don't get a good treatment  (it's using some old code path, which was never fixed by https://openjdk.org/jeps/416 too!), but with `MethodHandle` this is not true!

**Comment by chrisvest** [comment]
> @chrisvest The allocate uninitialized arrays method, if uses `Methodhandle::invokeExact` doesn't need to have _any_ limit: it will just go faster in any case. The limit in bytes was because reflective calls over intrinsic don't get a good treatment (it's using some old code path, which was never fixed by https://openjdk.org/jeps/416 too!), but with `MethodHandle` this is not true!

Oh, you mean we can remove the `UNINITIALIZED_ARRAY_ALLOCATION_THRESHOLD` checks?

> Additionally, I think we can target the `private` method instead of the public one -> https://github.com/openjdk/jdk/blob/9347bb7df845ee465c378c6f511ef8a6caea18ea/src/java.base/share/classes/jdk/internal/misc/Unsafe.java#L1385

The `PlatformDependent` method is public, so I think it's best if we keep the input validation checks.

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java (hunk 1 above)
Comment (chrisvest) [reply common/src/main/java/io/netty/util/internal/CleanerJava6.java:84]: Profiling shows the method handles add less than 1% overhead on top of the `Cleaner.clean()` call, so I think we're good.

**Comment by chrisvest** [comment]
Lot's of failures in `EpollSocketHalfClosedTest.testHalfClosureOnlyOneEventWhenAutoRead`, so that needs to be looked at.

**Comment by chrisvest** [comment]
The `EpollSocketHalfClosedTest.testHalfClosureOnlyOneEventWhenAutoRead` failures are not related to the changes in this PR, so I'm not letting that block merging.

**Review by normanmaurer — APPROVED** [review]

**Pull request merged**